clients/src/main/java/org/apache/kafka/common/KafkaFuture.java

```diff
@@ -101,10 +101,27 @@ private void maybeComplete() {
     /**
      * Returns a new KafkaFuture that, when this future completes normally, is executed with this
      * futures's result as the argument to the supplied function.
+     *
+     * The function may be called by the thread that calls {@code thenApply} or it may be called
+     * by the thread that completes this future.
+     *
+     * Exceptions thrown by the function will cause the resulting future to complete exceptionally,
+     * with the exception thrown by the function.
      */
     public abstract <R> KafkaFuture<R> thenApply(Function<T, R> function);
 
-    protected abstract void addWaiter(BiConsumer<? super T, ? super Throwable> action);
+    /**
+     * Add an action to be invoked on completion of this future (either normally or exceptionally)
+     *
+     * The action may be invoked by the thread that calls {@code addWaiter} or it may be invoked
+     * by the thread that completes the future.
+     *
+     * Exceptions thrown by the action will be caught and logged to prevent them from affecting
+     * other listeners or the calling thread.
```

Comment:
I don't agree with this.  If an exception is thrown, all the waiters should get it.

Also, I don't see why exceptions should be logged.  If the user wants to log the exception, they can do that in their own code.  The KafkaFuture code (like CompletableFuture) is extremely generic, so we don't know whether our users expect exceptions or not.

Reply:
@cmccabe We cannot guarantee that all waiters will receive the exception, since some waiters may already have accepted the result of the future by the time an exception is thrown.

Currently if an exception is thrown by a waiter it would prevent the other ones from getting invoked, that's why I'm proposing this solution. This is similar to what guava does when you add a listener to a ListenableFuture.

I don't believe a waiter should affect the outcome of a future unless we compose the existing future with the waiter to create a new future. If the underlying future succeeds, future.get() should always succeed regardless of what happens in any waiter.